### PR TITLE
105 product crud

### PR DESCRIPTION
**Pull request by MuhammadUsama1204**
Add a new product using a multi-steps form. Product should have a title, description, display picture, price, total quantity, and quantity in stock fields.
Edit an existing product.
Delete a product.
List all products.
Give the ability to filter out products on the basis of price and quantity.
Add styling

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,3 +1,11 @@
...
-    before_action :authenticate_user!
+
+  before_action :authenticate_user!
+
+  protected
+
+  def  after_sign_in_path_for(resource)
+    products_path
+  end
+
```
Comment (DaniyalXprolabs) [inline app/controllers/application_controller.rb]: Remove this extra line

File: app/controllers/products_controller.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  def destroy
+    @product = Product.find(params[:id])
+        if @product.destroy
+            redirect_to products_path, notice: "Product was successfully destroyed."
+        else
+            redirect_to products_path, alert: "Unable to destroy Product."
+        end
+    end
+
+  private
```
Comment (DaniyalXprolabs) [inline app/controllers/products_controller.rb:43]: Add an empty line here

File: app/views/products/_showform.html.erb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                      turbo_method: :delete,
+                      turbo_confirm: "Are you sure?"
+                    } %>
+      </th>
+    </tr>
+  </tbody>
+</table>
+
+<% if @product.display_picture.attached? %>
+Product Image<br><%= image_tag @product.display_picture, width: 500, height: 350, class: "product-image" %>
```
Comment (DaniyalXprolabs) [inline app/views/products/_showform.html.erb]: Fix the indentation issue here

File: config/environments/development.rb [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 Rails.application.configure do
   # Settings specified here will take precedence over those in config/application.rb.
-
+config.active_storage.variant_processor = :mini_magick
   # In the development environment your application's code is reloaded any time
```
Comment (DaniyalXprolabs) [inline config/environments/development.rb:6]: Fix the indentation issue here

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,3 +1,11 @@
 class ApplicationController < ActionController::Base
-    before_action :authenticate_user!
+
+  before_action :authenticate_user!
+
+  protected
+
+  def  after_sign_in_path_for(resource)
```
Comment (tauqeer-work92) [inline app/controllers/application_controller.rb]: remove extra space

File: app/controllers/products_controller.rb [hunk 6]
```diff
@@ -0,0 +1,76 @@
+class ProductsController < ApplicationController
+
+  def index
+      @products= Product.all
```
Comment (tauqeer-work92) [inline app/controllers/products_controller.rb]: use before action

@product = Product.find(params[:id])

File: app/controllers/users/sessions_controller.rb [hunk 7]
```diff
@@ -1,9 +1,6 @@
-# frozen_string_literal: true
```
Comment (tauqeer-work92) [inline app/controllers/users/sessions_controller.rb]: why remove

# frozen_string_literal: true

File: app/controllers/users/sessions_controller.rb [hunk 8]
```diff
@@ -1,9 +1,6 @@
...
-
 class Users::SessionsController < Devise::SessionsController
   before_action :configure_sign_in_params, only: [:create]
   protected
```
Comment (tauqeer-work92) [inline app/controllers/users/sessions_controller.rb:6]: line after protected

File: app/helpers/products_helper.rb [hunk 9]
```diff
@@ -0,0 +1,2 @@
+module ProductsHelper
```
Comment (tauqeer-work92) [inline app/helpers/products_helper.rb:1]: remove this file if not using it

File: app/views/products/_showform.html.erb [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+      <th scope="col">Description</th>
+      <th scope="col">quantity_in_stock</th>
+      <th scope="col">Price</th>
+      <th scope="col">Action</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th>
+        <%= @product.id%>
```
Comment (tauqeer-work92) [inline app/views/products/_showform.html.erb]: spacing issue in ruby tags

File: app/views/products/_showform.html.erb [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        <%= @product.title%>
+      </th>
+      <th>
+        <%= @product.description%>
+      </th>
+        <th>
+        <%= @product.quantity_in_stock%>
+      </th>
+      <th>
+        <%= @product.price%>
```
Comment (tauqeer-work92) [inline app/views/products/_showform.html.erb]: spacing issue in ruby tags

File: test/fixtures/products.yml [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
+  title: MyString
```
Comment (tauqeer-work92) [inline test/fixtures/products.yml]: remove this file.

File: test/models/product_test.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
```
Comment (tauqeer-work92) [inline test/models/product_test.rb:2]: remove this file.

File: Gemfile [hunk 14]
```diff
@@ -46,7 +46,7 @@ gem "bootsnap", require: false
 # gem "sassc-rails"
 
 # Use Active Storage variants [https://guides.rubyonrails.org/active_storage_overview.html#transforming-images]
-# gem "image_processing", "~> 1.2"
+gem "image_processing", "~> 1.2"
```
Comment (burraqxprolabs) [inline Gemfile]: I guess we don't have to need this gem

File: app/controllers/products_controller.rb [hunk 15]
```diff
@@ -0,0 +1,78 @@
+class ProductsController < ApplicationController
+  before_action :find_product, only: [:show, :edit, :update, :destroy]
+
+  def index
+      @products= Product.all
```
Comment (burraqxprolabs) [inline app/controllers/products_controller.rb]: Use tab spaces 2

File: app/controllers/products_controller.rb [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+      searchFilter
+  end
+
+  def show
```
Comment (burraqxprolabs) [inline app/controllers/products_controller.rb]: Def show; end

Use this syntax when we're not adding any code in any method

File: app/controllers/products_controller.rb [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    @product = Product.new(product_params)
+    if @product.save
+        flash[:success] = "Successfully Added"
+        redirect_to products_path
+    else
+        render :new, status: :unprocessable_entity
+    end
+  end
+
+  def edit
```
Comment (burraqxprolabs) [inline app/controllers/products_controller.rb]: Same here

File: app/controllers/products_controller.rb [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+  def product_params
+      params.require(:product).permit(:title, :description, :price, :total_quantity, :quantity_in_stock,:display_picture)
+  end
+
+  def searchFilter
+    if params[:search].present?
+      @products = @products.where("title ILIKE ?", "%#{params[:search]}%")
+    end
+
+    if params[:min_price].present? && params[:max_price].present?
```
Comment (burraqxprolabs) [inline app/controllers/products_controller.rb]: I Guess params[:min_price].present? || params[:max_price].present? we can use this condition here and we don't need these two methods

if params[:min_price].present?
      min_price = params[:min_price]
      @products = @products.where("price >= ?", min_price)
    end

    if params[:max_price].present?
      max_price = params[:max_price]
      @products = @products.where("price <= ?", max_price)
    end

File: app/controllers/users/sessions_controller.rb [hunk 19]
```diff
@@ -1,9 +1,10 @@
-# frozen_string_literal: true
+  # frozen_string_literal: true
```
Comment (burraqxprolabs) [inline app/controllers/users/sessions_controller.rb]: Please remove this space

File: app/controllers/products_controller.rb [hunk 20]
```diff
@@ -0,0 +1,53 @@
+class ProductsController < ApplicationController
+  before_action :find_product, only: [:show, :edit, :update, :destroy]
```
Comment (DaniyalXprolabs) [inline app/controllers/products_controller.rb]: %i[edit update show destroy]

File: app/views/products/edit.html.erb [hunk 21]
```diff
@@ -0,0 +1,2 @@
+<h1>Edit Product</h1>
+<%= render "form", product: @product%>
```
Comment (DaniyalXprolabs) [inline app/views/products/edit.html.erb]: Add an empty space after @product

File: app/views/products/index.html.erb [hunk 22]
```diff
@@ -0,0 +1 @@
+<%= render "formcard", product: @product  %>
```
Comment (DaniyalXprolabs) [inline app/views/products/index.html.erb:1]: Remove the 2 space at the end of render line

File: app/views/products/new.html.erb [hunk 23]
```diff
@@ -0,0 +1,2 @@
+<h1>New Product</h1>
+<%= render "form", product: @product%>
```
Comment (DaniyalXprolabs) [inline app/views/products/new.html.erb]: Add an empty space after @Product

File: app/views/products/show.html.erb [hunk 24]
```diff
@@ -0,0 +1 @@
+<%= render "showform", product: @product%>
```
Comment (DaniyalXprolabs) [inline app/views/products/show.html.erb]: Add an empty space after @Product

**Pull request merged**